torch/utils/checkpoint.py

```diff
@@ -0,0 +1,119 @@
+import torch
+
+
+def detach_variable(inputs):
+    if torch.is_tensor(inputs):
+        inp = inputs.detach()
+        return inp
+    elif isinstance(inputs, tuple):
+        return tuple(detach_variable(v) for v in inputs)
+    else:
+        raise RuntimeError("Unsupported input type: ", type(inputs).__name__)
+
+
+class CheckpointFunction(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        with torch.no_grad():
+            outputs = run_function(*args)
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *args):
+        assert (torch.autograd.is_checkpoint_valid()), \
+            "Checkpointing is not compatible with .grad(), please use .backward() if possible"
+        inputs = ctx.saved_tensors
+        inputs_list = detach_variable(inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs_list)
+
+        if isinstance(outputs, tuple):
+            output_list = list(outputs)
+        elif torch.is_tensor(outputs):
+            output_list = [outputs]
+        out_grads = [grad for grad in args]
+        torch.autograd.backward(output_list, out_grads)
+
+        input_grads = None
+        if isinstance(inputs_list, tuple):
+            input_grads = tuple(inp.grad for inp in inputs_list)
+            return (None,) + input_grads
+        elif torch.is_tensor(inputs_list):
+            input_grads = inputs_list.grad
+            return None, input_grads
+
+
+def checkpoint(run_function, *args):
+    r"""Checkpoint a model or part of the model
+
+    Checkpoint works by trading compute for memory. It can be applied on any
+    part of the model. In the forward pass, the model is run in volatile
+    manner i.e. the activations are not stored. The forward pass save the
+    inputs tuple and the run_function parameter. In the backwards pass, the
+    saved inputs and run_function is retreived, and the forward pass is done
+    on the model again (non-volatile this time) since we need to get the
+    activations values for calculating the gradient and then the gradients are
+    calculated.
+
+    Args:
+        run_function : describes what to run in the forward pass of the model or
+                       part of the model. It should also know how to handle
+                       the inputs passed as the tuple. For example, in LSTM,
+                       user passes (activation, hidden), run_function should
+                       correctly use first input as activation and second input
+                       as hidden
+        args:         tuple containing inputs to the run_function
+
+    Returns:
+        Output of running the run_function on *args
+    """
+    return CheckpointFunction.apply(run_function, *args)
+
+
+def checkpoint_sequential(modules, segments, *inputs):
```

Comment:
can you elaborate a bit more on what does "generic callables to be composed sequentially means"?

**For a bit of context**: These modules are left to user to pass and the modules themselves can be sequentials. We leave this to users so they can determine the granularity at which they want to checkpoint model. For example: in densenet, granularity can be denseblock or denselayer or something else. It highly depends on how the model is constructed by user. 

**NOTE:** This is a *convenience* method that's applicable to sequential type models like resnet, densenet etc. but this is not the only way for checkpointing. Check this out https://github.com/prigoyal/pytorch_memonger/blob/master/tutorial/Checkpointing_for_PyTorch_models.ipynb for various ways.

Reply:
I mean that there's nothing in this code that would require things in this list to be actual `nn.Module` instances. I know they *can* be, I'm saying that we could probably relax this constraint to generic callables (e.g. regular Python functions).

Reply:
got it. I think we can but I couldn't find a use case for this in various different types of models I have trained. Is this a blocking thing? I don't feel its a required feature based on my experience training models. :)

Reply:
It's not blocking, but:
1. It affects the interface, because the argument name could be chosen differently if we will ever expand support to callables. Argument names are part of the function interface in Python, so we can't change them at will.
2. It's purely a cosmetic change. Your code as is would work even if it was a list of non-module things. You only need to adjust the arg name (how about `steps` or `functions`?) and the docs!